storage/disk/paths.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2021 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package disk
+
+import (
+	"fmt"
+	"sort"
+
+	"github.com/open-policy-agent/opa/storage"
+)
+
+type pathMapper struct {
+	schemaVersion    string
+	partitionVersion string
+}
+
+func (pm *pathMapper) PolicyKey2ID(key []byte) string {
+	return string(key[len("/"+pm.schemaVersion+"/"+pm.partitionVersion+"/policies/"):])
+}
+
+func (pm *pathMapper) PolicyIDPrefix() []byte {
+	return []byte("/" + pm.schemaVersion + "/" + pm.partitionVersion + "/policies/")
+}
+
+func (pm *pathMapper) PolicyID2Key(id string) []byte {
+	return []byte("/" + pm.schemaVersion + "/" + pm.partitionVersion + "/policies/" + id)
+}
+
+func (pm *pathMapper) DataKey2Path(key []byte) (storage.Path, error) {
+	p, ok := storage.ParsePath(string(key))
+	if !ok {
+		return nil, &storage.Error{Code: storage.InternalErr, Message: fmt.Sprintf("corrupt key: %s", key)}
+	}
+	// skip /<schema_version>/<partition_version>/<data prefix>
+	return p[3:], nil
+}
+
+func (pm *pathMapper) DataPrefix2Key(path storage.Path) ([]byte, error) {
+	if len(path) == 0 {
+		return []byte("/" + pm.schemaVersion + "/" + pm.partitionVersion + "/data/"), nil
+	}
+	return []byte("/" + pm.schemaVersion + "/" + pm.partitionVersion + "/data" + path.String() + "/"), nil
+}
+
+func (pm *pathMapper) DataPath2Key(path storage.Path) ([]byte, error) {
+	if len(path) == 0 {
+		return nil, &storage.Error{Code: storage.InternalErr, Message: "empty path"}
+	}
+	return []byte("/" + pm.schemaVersion + "/" + pm.partitionVersion + "/data" + path.String()), nil
+}
+
+type pathSet []storage.Path
+
+func (ps pathSet) IsDisjoint() bool {
+	for i := range ps {
+		for j := range ps {
+			if i != j {
+				if ps[i].HasPrefix(ps[j]) || ps[j].HasPrefix(ps[i]) {
+					return false
+				}
+			}
+		}
+	}
+	return true
+}
```

Comment:
💭 Empty sets are disjoint from all sets then?

Reply:
right, the intersection of an empty path set and a non-empty path set is always empty.